README.md
# SystemPay payment acquirer

[![pipeline status](https://github.com/Horanet/payment_systempay/badges/11.0/pipeline.svg)](https://github.com/Horanet/payment_systempay/commits/11.0)
[![coverage report](https://github.com/Horanet/payment_systempay/badges/11.0/coverage.svg)](https://github.com/Horanet/payment_systempay/commits/11.0)

## Install Odoo 11

- [Install Odoo 11.0](https://www.odoo.com/documentation/11.0/setup/install.html) on your favorite environment

## Install addon via pip:

`pip install -e git+https://github.com/Horanet/payment_systempay.git@11.0#egg=odoo11-addon-payment-systempay`


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added

## [11.0.1.0.1] 2020-05-12
### Added
 - Set systempay_form as view_template_id for all existing systempay acquirers

## [11.0.1.0.0] - 2020-04-02
### Added
- Migration v11

[11.0.1.0.1]: https://github.com/Horanet/payment_systempay/compare/11.0.1.0.0...11.0.1.0.1
[Unreleased]: https://github.com/Horanet/payment_systempay/compare/11.0.1.0.1...dev-11.0

payment_systempay/tests/__init__.py
from . import test_systempay


odoo/addons/payment_systempay/tests/__init__.py
# coding: utf8

from . import test_systempay


payment_systempay/__manifest__.py
{
    # Module name in English
    'name': 'SystemPay Payment Acquirer',
    # Version, "odoo.min.yy.m.d"
    'version': '11.0.1.0.1',
    # Short description (with keywords)
    'summary': 'Payment Acquirer: SystemPay Implementation',
    # Description with metadata (in french)
    'description': """SystemPay Payment Acquirer""",
    'author': "Horanet",
    'website': "http://www.horanet.com/",
    # distribution license for the module (defaults: AGPL-3)
    'license': "AGPL-3",
    # Categories can be used to filter modules in modules listing. For the full list :
    # Check https://github.com/odoo/odoo/blob/master/openerp/addons/base/module/module_data.xml
    'category': 'Accounting',
    #
    'external_dependencies': {
        'python': []
    },
    # any module necessary for this one to work correctly. Either because this module uses features
    # they create or because it alters resources they define.
    'depends': [
        # --- Odoo --- #
        'payment'
        # --- External --- #

        # --- Horanet --- #
    ],
    # always loaded
    'css': [],
    'qweb': [],
    # list of XML files with data that will load to DB at moment when you install module
    'init_xml': [],
    # list of XML files with data that will load to DB at moment when you install or update module.
    'update_xml': [],
    # List of data files which must always be installed or updated with the module. A list of paths from the module root directory
    'data': [
        'views/payment_views.xml',
        'views/payment_systempay_templates.xml',

        'data/payment_acquirer.xml',
    ],
    # only loaded in demonstration mode
    'demo': [],
    'application': False,
    'auto_install': False,
    # permet d'installer automatiquement le module si toutes ses dépendances sont installés
    # -default value set is False
    # -If false, the dependent modules are not installed if not installed prior to the dependent module.
    # -If True, all corresponding dependent modules are installed at the time of installing this module.
    'installable': True,
    # -True, module can be installed.
    # -False, module is listed in application, but cannot install them.

    'post_init_hook': 'post_init_hook'
}


payment_systempay/models/inherited_payment_acquirer.py
import hashlib
import urllib.parse
import uuid
from datetime import datetime

from odoo import api, fields, models
from odoo.tools import float_round


class SystemPayAcquirer(models.Model):
    _inherit = 'payment.acquirer'

    provider = fields.Selection(selection_add=[('systempay', "SystemPay")])

    systempay_shop_id = fields.Char(string="Shop ID", required_if_provider='systempay')
    systempay_test_cert = fields.Char(string="Test certificate", required_if_provider='systempay')
    systempay_prod_cert = fields.Char(string="Prod certificate", required_if_provider='systempay')
    systempay_form_action_url = fields.Char(
        string="Form action URL",
        default="https://paiement.systempay.fr/vads-payment/",
        required_if_provider='systempay'
    )

    @api.model
    def _get_feature_support(self):
        """Get advanced feature support by provider.

        Each provider should add its technical in the corresponding
        key for the following features:
            * fees: support payment fees computations
            * authorize: support authorizing payment (separates
                         authorization and capture)
            * tokenize: support saving payment data in a payment.tokenize
                        object
        """
        res = super(SystemPayAcquirer, self)._get_feature_support()
        res['authorize'].append('paypal')
        return res

    @api.multi
    def systempay_generate_digital_sign(self, values):
        """Returns signature required by SystemPay to ensure integrity

        :param values: transaction values
        :return: sha1 hashed signature
        """
        self.ensure_one()

        systempay_values = {k: v for k, v in list(values.items()) if 'vads_' in k}
        data = []

        for _, value in sorted(systempay_values.items()):
            try:
                data.append(str(value))
            except UnicodeEncodeError:
                data.append(value)

        if self.environment == 'test':
            data.append(self.systempay_test_cert)
        elif self.environment == 'prod':
            data.append(self.systempay_prod_cert)

        signature = str('+').join(data)

        return hashlib.sha1(signature.encode('utf8')).hexdigest()

    @api.multi
    def systempay_get_form_action_url(self):
        self.ensure_one()

        return self.systempay_form_action_url

    @api.multi
    def systempay_form_generate_values(self, values):
        self.ensure_one()

        base_url = self.env['ir.config_parameter'].get_param('web.base.url')
        prec = self.env['decimal.precision'].precision_get('Product Price')

        mode = 'TEST'
        if self.environment == 'prod':
            mode = 'PRODUCTION'

        values = dict((k, v) for k, v in list(values.items()) if v)
        systempay_tx_values = dict(values)
        systempay_tx_values.update({
            'vads_site_id': self.systempay_shop_id,
            'vads_amount': int(float_round(values['amount'] * 100, prec)),
            'vads_currency': values.get('currency').number,
            'vads_trans_date': datetime.utcnow().strftime('%Y%m%d%H%M%S'),
            # SystemPay requires a unique 6-digits number between 000000 and 899999 per day
            'vads_trans_id': '%.6d' % int(uuid.uuid4().int % 899999),
            'vads_ctx_mode': mode,
            'vads_page_action': 'PAYMENT',
            'vads_action_mode': 'INTERACTIVE',
            'vads_payment_config': 'SINGLE',
            'vads_version': 'V2',
            'vads_return_mode': 'GET',
            'vads_url_return': '%s' % urllib.parse.urljoin(base_url, values.get('return_url')),
            'vads_order_id': values.get('reference').replace('/', ' '),

            'vads_cust_id': values.get('partner_id'),
            'vads_cust_first_name': values.get('partner_first_name', '')[0:62],
            'vads_cust_last_name': values.get('partner_last_name', '')[0:62],
            'vads_cust_address': values.get('partner_address', '')[0:254],
            'vads_cust_zip': values.get('partner_zip', '')[0:62],
            'vads_cust_city': values.get('partner_city', '')[0:62],
            'vads_cust_state': values.get('partner_state') and values.get('partner_state').name[0:127] or '',
            'vads_cust_country': values.get('partner_country') and values.get('partner_country').code or '',
            'vads_cust_email': values.get('partner_email', '')[0:126],
            'vads_cust_phone': values.get('partner_phone', '')[0:31],
        })

        systempay_tx_values['systempay_signature'] = self.systempay_generate_digital_sign(systempay_tx_values)

        return systempay_tx_values


payment_systempay/models/inherited_payment_transaction.py
import logging

from odoo import _, api, fields, models
from odoo.exceptions import ValidationError
from odoo.tools.float_utils import float_compare

_logger = logging.getLogger(__name__)

VADS_AUTH_RESULT = {
    '00': _("Approved or successfully processed transaction"),
    '02': _("Contact the card issuer"),
    '03': _("Invalid acceptor"),
    '04': _("Keep the card"),
    '05': _("Do not honor"),
    '07': _("Keep the card, special conditions"),
    '08': _("Confirm after identification"),
    '12': _("Invalid transaction"),
    '13': _("Invalid amount"),
    '14': _("Invalid cardholder number"),
    '15': _("Unknown issuer"),
    '17': _("Canceled by the buyer"),
    '19': _("Retry later"),
    '20': _("Incorrect response (error on the domain server)"),
    '24': _("Unsupported file update"),
    '25': _("Unable to locate the registered elements in the file"),
    '26': _("Duplicate registration, the previous record has been replaced"),
    '27': _("File update edit error"),
    '28': _("Denied access to file"),
    '29': _("Unable to update"),
    '30': _("Format error"),
    '31': _("Unknown acquirer company ID"),
    '33': _("Expired card"),
    '34': _("Fraud suspected"),
    '38': _("Expired card"),
    '41': _("Lost card"),
    '43': _("Stolen card"),
    '51': _("Insufficient balance or exceeded credit limit"),
    '54': _("Expired card"),
    '55': _("Invalid cardholder number"),
    '56': _("Card absent from the file"),
    '57': _("Transaction not allowed to this cardholder"),
    '58': _("Transaction not allowed to this cardholder"),
    '59': _("Suspected fraud"),
    '60': _("Card acceptor must contact the acquirer"),
    '61': _("Withdrawal limit exceeded"),
    '63': _("Security rules unfulfilled"),
    '68': _("Response not received or received too late"),
    '75': _("Number of attempts for entering the secret code has been exceeded"),
    '76': _("The cardholder is already blocked, the previous record has been saved"),
    '90': _("Temporary shutdown"),
    '91': _("Unable to reach the card issuer"),
    '94': _("Transaction duplicated"),
    '96': _("System malfunction"),
    '97': _("Overall monitoring timeout"),
    '98': _("Server not available, new network route requested"),
    '99': _("Initiator domain incident"),
    '000': _("Approved"),
    '001': _("Approve with ID"),
    '002': _("Partial Approval (Prepaid Cards only)"),
    '100': _("Declined"),
    '101': _("Expired Card / Invalid Expiration Date"),
    '106': _("Exceeded PIN attempts"),
    '107': _("Please Call Issuer"),
    '109': _("Invalid merchant"),
    '110': _("Invalid amount"),
    '111': _("Invalid account / Invalid MICR (Travelers Cheque)"),
    '115': _("Requested function not supported"),
    '117': _("Invalid PIN"),
    '119': _("Cardmember not enrolled / not permitted"),
    '122': _("Invalid card security code (a.k.a., CID, 4DBC, 4CSC)"),
    '125': _("Invalid effective date"),
    '181': _("Format error"),
    '183': _("Invalid currency code"),
    '187': _("Deny — New card issued"),
    '189': _("Deny — Account canceled"),
    '200': _("Deny — Pick up card"),
    '900': _("Accepted - ATC Synchronization"),
    '909': _("System malfunction (cryptographic error)"),
    '912': _("Issuer not available"),
}


class SystemPayTransaction(models.Model):
    _inherit = 'payment.transaction'

    @api.model
    def _systempay_form_get_tx_from_data(self, data):
        """ Method called by form_feedback after the transaction

        :param data: data received from the acquirer after the transaction
        :return: payment.transaction record if retrieved or an exception
        """

        reference = data.get('vads_order_id').replace(' ', '/')
        transactions = self.sudo().search([('reference', '=', reference)])

        if not transactions or len(transactions) > 1:
            error_msg = "SystemPay: received bad data for reference {}".format(reference)

            if not transactions:
                error_msg += "; no order found"
            else:
                error_msg += "; multiple order found"

            _logger.info(error_msg)
            raise ValidationError(error_msg)

        signature = data.get('signature')
        systempay_acquirer = self.env['payment.acquirer'].sudo().search([
            ('provider', '=', 'systempay')
        ])

        if signature != systempay_acquirer.systempay_generate_digital_sign(data):
            error_msg = _("SystemPay: signatures mismatch")
            _logger.info(error_msg)
            raise ValidationError(error_msg)

        return transactions

    @api.multi
    def _systempay_form_get_invalid_parameters(self, data):
        invalid_parameters = []

        if self.acquirer_reference and data.get('vads_trans_uuid') != self.acquirer_reference:
            invalid_parameters.append(
                ('vads_trans_uuid', data.get('vads_trans_uuid'), self.acquirer_reference)
            )

        if float_compare(float(data.get('vads_amount', '0.0')) / 100, (self.amount), 2) != 0:
            invalid_parameters.append(
                ('vads_amount', data.get('vads_amount'), '%.2f' % self.amount)
            )

        if self.partner_id.id and int(data.get('vads_cust_id')) != self.partner_id.id:
            invalid_parameters.append(
                ('vads_cust_id', data.get('vads_cust_id'), self.partner_id.id)
            )

        if self.acquirer_id and data.get('vads_site_id') != self.acquirer_id.systempay_shop_id:
            invalid_parameters.append(
                ('vads_shop_id', data.get('vads_site_id'), self.acquirer_id.systempay_shop_id)
            )

        return invalid_parameters

    @api.multi
    def _systempay_form_validate(self, data):
        """Check the status of the transaction and set it accordingly

        :param data: data received from systempay at the end of transaction
        """
        self.ensure_one()

        values = {
            'state_message': VADS_AUTH_RESULT.get(data.get('vads_auth_result')),
            'acquirer_reference': data.get('vads_trans_uuid')
        }

        transaction_status = data.get('vads_trans_status')

        if transaction_status == 'AUTHORISED':
            _logger.info("Validated SystemPay payment for transaction %s: set as done" % self.reference)
            values['state'] = 'done'
            values['date_validate'] = fields.Datetime.now()
        elif transaction_status == 'AUTHORISED_TO_VALIDATE':
            _logger.info("Validated SystemPay payment for transaction %s: set as done" % self.reference)
            values['state'] = 'authorized'
            values['date_validate'] = fields.Datetime.now()
        elif transaction_status == 'ABANDONED':
            _logger.info("Validated SystemPay payment for transaction %s: set as cancelled" % self.reference)
            values['state'] = 'cancel'
        elif transaction_status == 'INITIAL':
            _logger.info("Validated SystemPay payment for transaction %s: set as pending" % self.reference)
            values['state'] = 'pending'
        else:
            _logger.info("Validated SystemPay payment for transaction %s: set as error" % self.reference)
            values['state'] = 'error'

        return self.write(values)
